File: presets.py
from collections.abc import Callable
from typing import NamedTuple

import numpy as np


class PresetInfo(NamedTuple):
    name: str
    k_label: str = "Parameter:"
    k_decimals: int = 2
    k_step: float = 0.1
    k_min: float = -1000.0
    k_max: float = 1000.0
    k_default: float | Callable[[int], float] = 1.0

    p2_label: str | None = None
    p2_decimals: int = 0
    p2_step: float = 1.0
    p2_min: float = 1.0
    p2_max: float = 1000.0
    p2_default: float | Callable[[int], float] = 1.0

    p3_label: str | None = None
    p3_decimals: int = 2
    p3_step: float = 0.1
    p3_min: float = -np.pi
    p3_max: float = np.pi
    p3_default: float | Callable[[int], float] = 0.0


PRESETS = [
    PresetInfo(name="Random Angles"),
    PresetInfo(name="Twisted", k_label="Winding (k):", k_decimals=0, k_step=1.0, k_default=1.0),
    PresetInfo(name="Domain Wall"),
    PresetInfo(
        name="Vortex Band",
        k_label="Wraps (k):",
        k_decimals=0,
        k_step=1.0,
        k_default=1.0,
        p2_label="Width (w):",
        p2_default=1.0,
        p3_label="Shift (\u03b4\u03c6):",
        p3_default=0.0,
    ),
    PresetInfo(name="Cross Domain"),
    PresetInfo(
        name="Vortex Pair",
        k_label="Separation:",
        k_decimals=1,
        k_step=1.0,
        k_default=lambda l_side: float(l_side // 2),
    ),
    PresetInfo(
        name="Skyrmion",
        k_label="Radius (\u03c3):",
        k_decimals=1,
        k_step=1.0,
        k_default=lambda l_side: max(2.0, float(l_side) / 5.0),
    ),
    PresetInfo(
        name="Single Kick",
        k_label="Velocity (\u03c9):",
        k_decimals=2,
        k_step=0.1,
        k_default=5.0,
    ),
    PresetInfo(
        name="Thermalized",
        k_label="Mean Energy (\u03b5):",
        k_decimals=2,
        k_step=0.1,
        k_default=1.0,
        k_min=0.0,
    ),
]


def get_preset_by_name(name: str) -> PresetInfo:
    for p in PRESETS:
        if p.name == name:
            return p
    return PRESETS[0]


def generate_initial_state(
    l_side: int, preset_name: str, k: float, p2: float, p3: float, temp: float
) -> np.ndarray:
    """Generate initial state based on the selected preset and parameters."""
    n = l_side**2
    y0 = np.zeros(2 * n)

    if preset_name == "Random Angles":
        # theta_i from [-pi, pi)
        y0[:n] = np.random.uniform(-np.pi, np.pi, n)
    elif preset_name == "Twisted":
        # theta_i,j = 2*pi*k*i/L (twist along x)
        i_indices = np.arange(l_side).repeat(l_side).reshape(l_side, l_side).T.flatten()
        y0[:n] = (2 * np.pi * k * i_indices) / l_side
    elif preset_name == "Domain Wall":
        # Half at 0, half at pi (split along x)
        theta_2d = np.zeros((l_side, l_side))
        half = l_side // 2
        theta_2d[half:, :] = np.pi
        y0[:n] = theta_2d.flatten()
        # Tiny velocity perturbation to break unstable equilibrium
        y0[n] = 1e-6
    elif preset_name == "Vortex Band":
        # A vertical band of phase ramps
        theta_2d = np.zeros((l_side, l_side))
        w = int(p2)
        delta_phi = p3

        mid = l_side // 2
        start = max(0, mid - w // 2)
        end = min(l_side, start + w)

        # Phase ramp along y
        ramp = np.linspace(0, 2 * np.pi * k, l_side, endpoint=False)

        for j in range(start, end):
            # Apply ramp and inter-line phase shift
            theta_2d[:, j] = ramp + (j - start) * delta_phi

        y0[:n] = theta_2d.flatten()
    elif preset_name == "Cross Domain":
        # Four triangular domains (Upper/Lower = pi/2, Left/Right = -pi/2)
        theta_2d = np.zeros((l_side, l_side))
        yy, xx = np.indices((l_side, l_side))
        # Diagonals: y=x and y=L-1-x
        mask_upper = (yy < xx) & (yy < (l_side - 1 - xx))
        mask_lower = (yy > xx) & (yy > (l_side - 1 - xx))
        mask_left = (yy > xx) & (yy < (l_side - 1 - xx))
        mask_right = (yy < xx) & (yy > (l_side - 1 - xx))

        theta_2d[mask_upper] = np.pi / 2
        theta_2d[mask_lower] = np.pi / 2
        theta_2d[mask_left] = -np.pi / 2
        theta_2d[mask_right] = -np.pi / 2
        y0[:n] = theta_2d.flatten()
    elif preset_name == "Vortex Pair":
        # Two opposite vortices
        yy, xx = np.indices((l_side, l_side))
        mid_f = (l_side - 1) / 2.0
        sep = k / 2.0

        # Vortex at (mid - sep, mid), Antivortex at (mid + sep, mid)
        v1 = np.arctan2(yy - mid_f, xx - (mid_f - sep))
        v2 = np.arctan2(yy - mid_f, xx - (mid_f + sep))
        y0[:n] = (v1 - v2).flatten()
    elif preset_name == "Skyrmion":
        # Localized phase twist
        yy, xx = np.indices((l_side, l_side))
        mid_f = (l_side - 1) / 2.0
        sigma_f = float(k)
        r_sq = (xx - mid_f) ** 2 + (yy - mid_f) ** 2
        y0[:n] = (np.pi * np.exp(-r_sq / (2 * sigma_f**2))).flatten()
    elif preset_name == "Single Kick":
        # Gaussian velocity kick (Wave Packet)
        yy, xx = np.indices((l_side, l_side))
        mid_f = (l_side - 1) / 2.0
        omega_peak = k
        # Fixed width for the kick "drop"
        sigma_f = 2.0
        r_sq = (xx - mid_f) ** 2 + (yy - mid_f) ** 2
        kick = omega_peak * np.exp(-r_sq / (2 * sigma_f**2))
        y0[n:] = kick.flatten()
    elif preset_name == "Thermalized":
        # Random velocities (Maxwell-Boltzmann like)
        sigma_f = np.sqrt(max(0, 2 * k))
        y0[n:] = np.random.normal(0, sigma_f, n)

    # Add Thermal Noise Overlay (Phonons)
    if temp > 0:
        # sigma = sqrt(2 * T)
        noise_sigma = np.sqrt(2.0 * temp)
        y0[n:] += np.random.normal(0, noise_sigma, n)

    # Wrap theta to [-pi, pi)
    y0[:n] = (y0[:n] + np.pi) % (2 * np.pi) - np.pi

    return y0

File: web/src/presets.ts
export interface PresetInfo {
  name: string;
  kLabel?: string;
  kDecimals: number;
  kStep: number;
  kMin: number;
  kMax: number;
  kDefault: number | ((l: number) => number);

  p2Label?: string;
  p2Decimals: number;
  p2Step: number;
  p2Min: number;
  p2Max: number;
  p2Default: number | ((l: number) => number);

  p3Label?: string;
  p3Decimals: number;
  p3Step: number;
  p3Min: number;
  p3Max: number;
  p3Default: number | ((l: number) => number);
}

const DEFAULT_PRESET: PresetInfo = {
  name: "Default",
  kLabel: "Parameter:",
  kDecimals: 2,
  kStep: 0.1,
  kMin: -1000,
  kMax: 1000,
  kDefault: 1.0,
  p2Decimals: 0,
  p2Step: 1,
  p2Min: 1,
  p2Max: 1000,
  p2Default: 1,
  p3Decimals: 2,
  p3Step: 0.1,
  p3Min: -Math.PI,
  p3Max: Math.PI,
  p3Default: 0,
};

function createPreset(base: Partial<PresetInfo>): PresetInfo {
  return { ...DEFAULT_PRESET, ...base };
}

export const PRESETS: PresetInfo[] = [
  createPreset({ name: "Random Angles" }),
  createPreset({
    name: "Twisted",
    kLabel: "Winding (k):",
    kDecimals: 0,
    kStep: 1.0,
    kDefault: 1.0,
  }),
  createPreset({ name: "Domain Wall" }),
  createPreset({ name: "Pi/2 Domain Wall" }),
  createPreset({
    name: "Vortex Band",
    kLabel: "Wraps (k):",
    kDecimals: 0,
    kStep: 1.0,
    kDefault: 1.0,
    p2Label: "Width (w):",
    p2Default: 1.0,
    p3Label: "Shift (δφ):",
    p3Default: 0.0,
  }),
  createPreset({ name: "Cross Domain" }),
  createPreset({
    name: "Vortex Pair",
    kLabel: "Separation:",
    kDecimals: 1,
    kStep: 1.0,
    kDefault: (l) => Math.floor(l / 2),
  }),
  createPreset({
    name: "Skyrmion",
    kLabel: "Radius (σ):",
    kDecimals: 1,
    kStep: 1.0,
    kDefault: (l) => Math.max(2.0, l / 5.0),
  }),
  createPreset({
    name: "Single Kick",
    kLabel: "Velocity (ω):",
    kDecimals: 2,
    kStep: 0.1,
    kDefault: 5.0,
  }),
  createPreset({
    name: "Thermalized",
    kLabel: "Mean Energy (ε):",
    kDecimals: 2,
    kStep: 0.1,
    kDefault: 1.0,
    kMin: 0.0,
  }),
];

export function getPresetByName(name: string): PresetInfo {
  return PRESETS.find((p) => p.name === name) || PRESETS[0];
}

export function generateInitialState(
  l: number,
  presetName: string,
  k: number,
  p2: number,
  p3: number,
  temp: number,
): { theta: Float64Array; omega: Float64Array } {
  const n = l * l;
  const theta = new Float64Array(n);
  const omega = new Float64Array(n);

  // Helper for 2D indices
  // row=y, col=x
  const getIdx = (r: number, c: number) => r * l + c;

  if (presetName === "Random Angles") {
    for (let i = 0; i < n; i++) {
      theta[i] = Math.random() * 2 * Math.PI - Math.PI;
    }
  } else if (presetName === "Twisted") {
    // theta_ij = 2*pi*k*i/L (twist along x/col or y/row?)
    // Python: i_indices from arange(l).repeat(l)... transpose...
    // indices in python: i increases down rows?
    // i_indices = [0, 0, ..., 1, 1, ... ] (l times) if repeated correctly
    // Python: np.arange(l).repeat(l).reshape(l, l).T.flatten()
    // T makes it vary along columns?
    // np.arange(l).repeat(l).reshape(l, l) -> [[0..0], [1..1]..] (rows constant)
    // .T -> [[0, 1, ..], [0, 1, ..]] (cols constant?)
    // Let's just implement Twist along X: theta(x,y) = 2*pi*k*x/L
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        theta[getIdx(r, c)] = (2 * Math.PI * k * r) / l; // Twist along rows (Y) like Python's i_indices usually
      }
    }
  } else if (presetName === "Domain Wall") {
    const half = Math.floor(l / 2);
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        theta[getIdx(r, c)] = r >= half ? Math.PI : 0;
      }
    }
    // Kick first rotor slightly
    omega[0] = 1e-6;
  } else if (presetName === "Pi/2 Domain Wall") {
    const half = Math.floor(l / 2);
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        theta[getIdx(r, c)] = r < half ? Math.PI / 2 : -Math.PI / 2;
      }
    }
    // Kick first rotor slightly
    omega[0] = 1e-6;
  } else if (presetName === "Vortex Band") {
    const w = Math.floor(p2);
    const deltaPhi = p3;
    const mid = Math.floor(l / 2);
    const start = Math.max(0, mid - Math.floor(w / 2));
    const end = Math.min(l, start + w);

    for (let c = start; c < end; c++) {
      for (let r = 0; r < l; r++) {
        const ramp = (2 * Math.PI * k * r) / l;
        theta[getIdx(r, c)] = ramp + (c - start) * deltaPhi;
      }
    }
  } else if (presetName === "Cross Domain") {
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        // y=r, x=c
        const upper = r < c && r < l - 1 - c;
        const lower = r > c && r > l - 1 - c;
        const left = r > c && r < l - 1 - c;
        const right = r < c && r > l - 1 - c;

        let val = 0;
        if (upper || lower) val = Math.PI / 2;
        if (left || right) val = -Math.PI / 2;
        theta[getIdx(r, c)] = val;
      }
    }
  } else if (presetName === "Vortex Pair") {
    const mid = (l - 1) / 2.0;
    const sep = k / 2.0;
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        // v1 at (mid - sep, mid) -- wait, x or y?
        // Python: v1 = arctan2(yy - mid, xx - (mid - sep))
        // yy is row index?
        const y = r;
        const x = c;
        const v1 = Math.atan2(y - mid, x - (mid - sep));
        const v2 = Math.atan2(y - mid, x - (mid + sep));
        theta[getIdx(r, c)] = v1 - v2;
      }
    }
  } else if (presetName === "Skyrmion") {
    const mid = (l - 1) / 2.0;
    const sigma = k;
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        const rSq = (c - mid) ** 2 + (r - mid) ** 2;
        theta[getIdx(r, c)] = Math.PI * Math.exp(-rSq / (2 * sigma * sigma));
      }
    }
  } else if (presetName === "Single Kick") {
    const mid = (l - 1) / 2.0;
    const omegaPeak = k;
    const sigma = 2.0;
    for (let r = 0; r < l; r++) {
      for (let c = 0; c < l; c++) {
        const rSq = (c - mid) ** 2 + (r - mid) ** 2;
        omega[getIdx(r, c)] = omegaPeak * Math.exp(-rSq / (2 * sigma * sigma));
      }
    }
  } else if (presetName === "Thermalized") {
    const sigma = Math.sqrt(Math.max(0, 2 * k));
    for (let i = 0; i < n; i++) {
      // Box-Muller transform for Gaussian
      const u1 = Math.random();
      const u2 = Math.random();
      const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      omega[i] = z * sigma;
    }
  }

  // Add thermal noise overlay
  if (temp > 0) {
    const noiseSigma = Math.sqrt(2.0 * temp);
    for (let i = 0; i < n; i++) {
      const u1 = Math.random();
      const u2 = Math.random();
      const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      omega[i] += z * noiseSigma;
    }
  }

  // Wrap theta
  for (let i = 0; i < n; i++) {
    let th = theta[i];
    th = (th + Math.PI) % (2 * Math.PI);
    if (th < 0) th += 2 * Math.PI;
    th -= Math.PI;
    theta[i] = th;
  }

  return { theta, omega };
}
